### ISSUE TITLE
CentOS 7 compilation failure

### ISSUE THREAD

**Issue by schwartz-je** [latchset/pkcs11-provider#338]
When compiling on an CentOS 7 host, I got a lot of errors like
      missing braces around initializer [-Werror=missing-braces]
As the compiliation flag -Werror is set, I can't compile without making corrections

```
$ ./configure --disable-static --disable-valgrind --prefix=/home/opensrc/local/openssl-3.0 OPENSSL_CFLAGS=-I/home/opensrc/local/openssl-3.0/include CFLAGS=-I/home/opensrc/local/include -m64 -O2 -Wa,--noexecstack -fPIC LDFLAGS=-L/home/opensrc/local/openssl-3.0/lib CRYPTO_LIBS=-Wl,-Bstatic -lcrypto -Wl,-Bdynamic CRYPTO_CFLAGS=-Ujes OPENSSL_LIBS=-lcrypto -lssl
$ make
encoder.c: In function 'p11prov_ec_pubkey_to_x509':
encoder.c:548:12: error: missing braces around initializer [-Werror=missing-brac
es]
     struct ecdsa_key_point keypoint = { 0 };
            ^
encoder.c:548:12: error: (near initialization for 'keypoint.curve') [-Werror=mis
sing-braces]
cc1: all warnings being treated as errors
make[2]: *** [pkcs11_la-encoder.lo] Erreur 1
make[1]: *** [all] Erreur 2 
make: *** [all-recursive] Erreur 1
```
Impacted files ares:

```
--- src/encoder.c.orig  2024-01-09 20:18:47.000000000 +0100
+++ src/encoder.c       2024-01-11 17:24:28.647444319 +0100
@@ -545,10 +545,11 @@
     
 static X509_PUBKEY *p11prov_ec_pubkey_to_x509(P11PROV_OBJ *key)
 {
-    struct ecdsa_key_point keypoint = { 0 };
+    struct ecdsa_key_point keypoint;
     X509_PUBKEY *pubkey;
     int ret;
 
+    memset (&keypoint, 0, sizeof (keypoint));
     ret = p11prov_obj_export_public_key(
         key, CKK_EC, true, p11prov_ec_set_keypoint_data, &keypoint);
     if (ret != RET_OSSL_OK) {
```

```
--- src/interface.c.orig        2024-01-09 20:18:47.000000000 +0100
+++ src/interface.c     2024-01-12 10:03:25.640258496 +0100
@@ -373,14 +373,16 @@
 
 CK_RV p11prov_module_reinit(P11PROV_MODULE *mctx)
 {
-    CK_C_INITIALIZE_ARGS args = { 0 };
-    CK_INFO ck_info = { 0 };
+    CK_C_INITIALIZE_ARGS args;
+    CK_INFO ck_info;
     CK_RV ret;
 
     if (!mctx) {
         return CKR_GENERAL_ERROR;
     }
 
+    memset (&args, 0, sizeof (args));
+    memset (&ck_info, 0, sizeof (ck_info));
     ret = MUTEX_LOCK(mctx);
     if (ret != CKR_
```
```
-- src/objects.c.orig  2024-01-09 20:18:47.000000000 +0100
+++ src/objects.c       2024-01-11 17:24:28.648444319 +0100
@@ -974,7 +974,7 @@
     CK_OBJECT_CLASS class = p11prov_uri_get_class(uri);
     CK_ATTRIBUTE id = p11prov_uri_get_id(uri);
     CK_ATTRIBUTE label = p11prov_uri_get_label(uri);
-    CK_ATTRIBUTE template[3] = { 0 };
+    CK_ATTRIBUTE template[3];
     CK_OBJECT_HANDLE *objects = NULL;
     CK_ULONG tsize = 0;
     CK_ULONG objcount = 0;
@@ -986,6 +986,7 @@
                   id.ulValueLen,
                   label.type == CKA_LABEL ? (char *)label.pValue : "None");

+    memset (template, 0, sizeof (template));
     switch (class) {
     case CKO_CERTIFICATE:
     case CKO_PUBLIC_KEY:
@@ -1075,7 +1076,7 @@
 static P11PROV_OBJ *find_associated_obj(P11PROV_CTX *provctx, P11PROV_OBJ *obj,
                                         CK_OBJECT_CLASS class)
 {
-    CK_ATTRIBUTE template[2] = { 0 };
+    CK_ATTRIBUTE template[2];
     CK_ATTRIBUTE *id;
     CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
     P11PROV_SESSION *session = NULL;
@@ -1087,6 +1088,7 @@

     P11PROV_debug("Find associated object");

+    memset (template, 0, sizeof (template));
     id = p11prov_obj_get_attr(obj, CKA_ID);
     if (!id || id->ulValueLen == 0) {
         P11PROV_raise(provctx, CKR_GENERAL_ERROR, "No CKA_ID in source object");
@@ -1147,7 +1149,7 @@
     CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
     P11PROV_SESSION *session = NULL;
     CK_SESSION_HANDLE sess = CK_INVALID_HANDLE;
-    CK_ATTRIBUTE template[3] = { 0 };
+    CK_ATTRIBUTE template[3];
     CK_ATTRIBUTE *attr;
     int anum;
     CK_OBJECT_HANDLE handle;
@@ -1178,6 +1180,7 @@

     sess = p11prov_session_handle(session);

+    memset (template, 0, sizeof(template));
     anum = 0;
     CKATTR_ASSIGN(template[anum], CKA_CLASS, &(obj->class), sizeof(obj->class));
     anum++;
@@ -1410,7 +1413,7 @@
 #define MAX_KEY_ATTRS 2
 static CK_RV get_all_from_cert(P11PROV_OBJ *crt, CK_ATTRIBUTE *attrs, int num)
 {
-    OSSL_PARAM params[MAX_KEY_ATTRS + 1] = { 0 };
+    OSSL_PARAM params[MAX_KEY_ATTRS + 1];
     CK_ATTRIBUTE_TYPE types[MAX_KEY_ATTRS];
     CK_ATTRIBUTE *type;
     CK_ATTRIBUTE *value;
@@ -1421,6 +1424,7 @@
     int ret;
     CK_RV rv;

+    memset (params, 0, sizeof (params));
     /* if CKA_CERTIFICATE_TYPE is not present assume CKC_X_509 */
     type = p11prov_obj_get_attr(crt, CKA_CERTIFICATE_TYPE);
     if (type) {
@@ -1657,11 +1661,12 @@
 static int p11prov_obj_export_public_rsa_key(P11PROV_OBJ *obj,
                                              OSSL_CALLBACK *cb_fn, void *cb_arg)
 {
-    CK_ATTRIBUTE attrs[RSA_PUB_ATTRS] = { 0 };
+    CK_ATTRIBUTE attrs[RSA_PUB_ATTRS];
     OSSL_PARAM params[RSA_PUB_ATTRS + 1];
     CK_RV rv;
     int ret;

+    memset (attrs, 0, sizeof (attrs));
     if (p11prov_obj_get_key_type(obj) != CKK_RSA) {
         return RET_OSSL_ERR;
     }
@@ -1850,8 +1855,8 @@
 static int p11prov_obj_export_public_ec_key(P11PROV_OBJ *obj,
                                             OSSL_CALLBACK *cb_fn, void *cb_arg)
 {
-    CK_ATTRIBUTE attrs[EC_MAX_PUB_ATTRS] = { 0 };
-    OSSL_PARAM params[EC_MAX_OSSL_PARAMS + 1] = { 0 };
+    CK_ATTRIBUTE attrs[EC_MAX_PUB_ATTRS];
+    OSSL_PARAM params[EC_MAX_OSSL_PARAMS + 1];
     CK_KEY_TYPE key_type;
     int nattr = 0;
     int nparam = 0;
@@ -1860,6 +1865,8 @@
     EC_GROUP *group = NULL;
     int curve_nid = NID_undef;

+    memset (attrs, 0, sizeof (attrs));
+    memset (params, 0, sizeof (params));
     key_type = p11prov_obj_get_key_type(obj);
     switch (key_type) {
     case CKK_EC:
```

```
--- src/provider.c.orig 2024-01-09 20:18:47.000000000 +0100
+++ src/provider.c      2024-01-12 10:08:18.684248940 +0100
@@ -618,7 +618,8 @@
 CK_INFO p11prov_ctx_get_ck_info(P11PROV_CTX *ctx)
 {
     if (!ctx->module) {
-        CK_INFO info = { 0 };
+        CK_INFO info;
+        memset (&info, 0, sizeof(info));
         return info;
     }
     return p11prov_module_ck_info(ctx->module);
```

**Comment by schwartz-je** [comment]
Sorry to put diff into the message itself, but when I tried to join a encoder.c.patch file, I got a github error:
  **We don't support that file type.
  try again with .....**

**Comment by schwartz-je** [comment]
Compiler is gcc 4.8.5

**Comment by simo5** [comment]
CentOS 7 is not a target for us TBH given it doesn't ship (and will never ship) OpenSSL 3.0
What makes you want to build specifically on that platform ?

**Comment by schwartz-je** [comment]
See reason in #337

**Comment by Jakuje** [comment]
I did some search and it looks like this is actually a 10 years fixed bug in gcc:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119

It should have been fixed somewhere around 4.9. Certainly I do not think we want to provide compatiblity for this bugs. If you need to use this old gcc, override CFLAGS with environment variable.
